File: 114. Flatten Binary Tree to Linked List.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <queue>
class Solution {
private:
    queue <TreeNode*> q; 
public:
    void flatten(TreeNode* root) {
        if (!(root == NULL || (root->left == NULL && root->right == NULL)))
        {
            TreeNode* p = root;
            preorder(p);
            while (!q.empty())
            {
                TreeNode* curr = q.front();
                q.pop();
                p->right = curr;
                p->left = NULL;
                p = p->right;
            }
        }
    }
    void preorder(TreeNode* p)
    {
        if (p == NULL) return;
        q.push(p);
        preorder(p->left);
        preorder(p->right);
    }
};

File: 114. Flatten Binary Tree to Linked List.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        if not (root == None or (root.left==None and root.right == None)):
            self.queue = []
            p = root
            self.preorder(p)
            while len(self.queue) > 0:
                curr = self.queue.pop(0)
                p.right = curr
                p.left = None
                p = p.right
        
    def preorder(self, p):
        if not p:
            return
        self.queue += [p]
        self.preorder(p.left)
        self.preorder(p.right)
